wagtail/contrib/settings/models.py

```diff
@@ -161,51 +219,84 @@ class Meta:
         abstract = True
 
     @classmethod
-    def _get_or_create(cls):
-        """
-        Internal convenience method to get or create the first instance.
-
-        We cannot hardcode `pk=1`, for example, as not all database backends
-        use sequential IDs (e.g. Postgres).
-        """
-
-        first_obj = cls.base_queryset().first()
-        if first_obj is None:
-            return cls.objects.create()
-        return first_obj
+    def _get_or_create(cls, **kwargs):
+        raise NotImplementedError
 
     @classmethod
-    def load(cls, request_or_site=None):
+    def load(cls, request_or_site=None, locale=None):
         """
-        Get or create an instance of this model. There is only ever one
-        instance of models inheriting from `AbstractSetting` so we can
-        use `pk=1`.
-
+        Get or create an instance of this model.
         If `request_or_site` is present and is a request object, then we cache
         the result on the request for faster repeat access.
         """
-
         # We can only cache on the request, so if there is no request then
         # we know there's nothing in the cache.
         if request_or_site is None or isinstance(request_or_site, Site):
-            return cls._get_or_create()
+            return cls._get_or_create(locale=locale)
 
         # Check if we already have this in the cache and return it if so.
         attr_name = cls.get_cache_attr_name()
         if hasattr(request_or_site, attr_name):
             return getattr(request_or_site, attr_name)
 
-        obj = cls._get_or_create()
+        locale = cls._get_locale(locale=locale, request=request_or_site)
+        obj = cls._get_or_create(locale=locale)
 
         # Cache for next time.
         setattr(request_or_site, attr_name, obj)
-
         return obj
 
     def __str__(self):
         return self._meta.verbose_name
 
 
+class BaseGenericSetting(AbstractGenericSetting):
+    class Meta:
+        abstract = True
+
+    @classmethod
+    def _get_or_create(cls, **kwargs):
+        """
+        Internal convenience method to get or create the first instance.
+
+        We cannot hardcode `pk=1`, for example, as not all database backends
+        use sequential IDs (e.g. Postgres).
+        """
+
+        first_obj = cls.base_queryset().first()
+        if first_obj is None:
+            return cls.objects.create()
+        return first_obj
+
+
+class BaseTranslatableGenericSetting(TranslatableMixin, AbstractGenericSetting):
```

Comment:
I feel that this adds too much developer overhead, we are not asking people to use `TranslatablePage` instead of `Page` if they want content to support i18n. And we do offer `TranslatableMixin` for models outside of our control, which is fine.

IMHO, as a developer I should only need to extend `BaseGenericSetting` or `BaseSiteSetting`, and if/when I add i18n, "it should just work" ™️.

Reply:
I copied the approach @kaedroho did originally to split the models between (non)/translatable settings. So maybe it'd be nice to know the rationale behind it. Otherwise, I agree with how it'd simplify things for developers.

Reply:
The reason for splitting them was because of the unique key on `site`. Introducing locales would mean we would need multiple settings per site, one for each locale.

Reply:
That would be the expected behaviour, though. No?
with `BaseTranslatableSiteSetting` you still end up with multiple settings per site:

- Site A, English
- Site A, French
- Site B, English
etc.

Reply:
Yep, but if you added `TranslatableMixin` to a `BaseSiteSetting`, the unique key on the `site` field would prevent Site A from having two settings objects

Reply:
I get that. What I am suggesting here is that we add `unique_together = [("site", "locale"), ("translation_key", "locale")]` to `BaseSiteSetting` as it is done for `BaseTranslatableSiteSetting`, thus removing the need for `BaseTranslatableSiteSetting`

Reply:
Should I go ahead and unify the classes then?

Reply:
My preference would be to unify them. @kaedroho?

Reply:
@zerolab With respect, this work was sponsored by a client, and it so happens that they have a combination of settings models... some of which require translation, and others that do not. For example, there are `DonationSettings` and `MarketingSignupSettings` that contain only configurational (non user-facing) things that do not (and should not) change on a per-language basis. There are others too, such as `SocialMediaLinks`, where the values ARE user-facing, but are simply a series of URLFields, where the same values apply for all locales. Another example would be GA settings.

In my view, Pages are a different kettle of fish entirely, so there's no need to emulate what happens there. Settings usage, in my experience, tends to be much similar to Snippets, where it's also perfectly valid to have both translatable and non-translatable models. Developers generally just use Settings over Snippets to get the 'singleton-like' interface and easy access in templates.

Reply:
@ababic I am less concerned about something being sponsored by a client per se as we want to make it work for everyone.

Thank you for the additional example -- yep, that is a real and valid use case that I did not have at the back of my mind last week. This is the type of detail PR descriptions could really do with.

Happy for them to be separate on that basis